src/apps/lwaftr/fragmentv4_hardened.lua

```diff
@@ -0,0 +1,305 @@
+
+local function get_frag_id(frag)
+   local o_id = ehs + o_ipv4_identification
+   return ntohs(rd16(frag.data + o_id))
+end
+
+-- The most significant three bits are other information, and the
+-- offset is expressed in 8-octet units, so just mask them off and * 8 it.
+local function get_frag_start(frag)
+   local o_fstart = ehs + o_ipv4_flags
+   local raw_start = ntohs(rd16(frag.data + o_fstart))
+   return band(raw_start, 0x1fff) * 8
+end
+
+-- This is the 'MF' bit of the IPv4 fragment header; it's the 3rd bit
+-- of the flags.
+local function is_last_fragment(frag)
+   local o_flag = ehs + o_ipv4_flags
+   return band(frag.data[o_flag], 0x20) == 0
+end
+
+local function get_key(fragment)
+   local key = scratch_fragkey
+   local o_src = ehs + o_ipv4_src_addr
+   local o_dst = ehs + o_ipv4_dst_addr
+   local o_id = ehs + o_ipv4_identification
+   ffi.copy(key.src_addr, fragment.data + o_src, 4)
+   ffi.copy(key.dst_addr, fragment.data + o_dst, 4)
+   key.fragment_id = ntohs(rd16(fragment.data + o_id))
+   return key
+end
+
+local function free_reassembly_buf_and_pkt(pkt, frags_table)
+   local key = get_key(pkt)
+   frags_table:remove(key, false)
+   packet.free(pkt)
+end
+
+local function swap(array, i, j)
+   local tmp = array[j]
+   array[j] = array[i]
+   array[i] = tmp
+end
+
+-- This is an insertion sort, and only called on 2+ element arrays
+local function sort_array(array, last_index)
+   for i=0,last_index do
+      local j = i
+      while j > 0 and array[j-1] > array[j] do
+         swap(array, j, j-1)
+         j = j - 1
+      end
+   end
+end
+
+local function verify_valid_offsets(reassembly_buf)
+   if reassembly_buf.fragment_starts[0] ~= 0 then
+      return false
+   end
+   for i=1,reassembly_buf.fragment_count-1 do
+      if reassembly_buf.fragment_starts[i] ~= reassembly_buf.fragment_ends[i-1] then
+         return false
+      end
+   end
+   return true
+end
+
+local function reassembly_status(reassembly_buf)
+   if reassembly_buf.final_start == 0 then
+      return FRAGMENT_MISSING
+   end
+   if reassembly_buf.running_length ~= reassembly_buf.reassembly_packet.length then
+      return FRAGMENT_MISSING
+   end
+   if not verify_valid_offsets(reassembly_buf) then
+      return REASSEMBLY_INVALID
+   end
+   return REASSEMBLY_OK
+end
+
+-- IPv4 requires recalculating an embedded checksum.
+local function fix_pkt_checksum(pkt)
+   local ihl = get_ihl_from_offset(pkt, ehs)
+   local checksum_offset = ehs + o_ipv4_checksum
+   wr16(pkt.data + checksum_offset, 0)
+   wr16(pkt.data + checksum_offset,
+        htons(ipsum(pkt.data + ehs, ihl, 0)))
+end
+
+local function attempt_reassembly(frags_table, reassembly_buf, fragment)
+   local ihl = get_ihl_from_offset(fragment, ehs)
+   local reassembly_pkt = reassembly_buf.reassembly_packet
+   local frag_id = get_frag_id(fragment)
+   if frag_id ~= reassembly_buf.fragment_id then -- unreachable
+      assert(false, "Impossible case reached in v4 reassembly") --REASSEMBLY_INVALID
```

Comment:
I think is clearer written as:

```
error("Impossible case reached in v4 reassembly")
```

I saw you changed it in the IPv6 counterpart, but left the first `false` argument. Need to remove it.
Same as `error("Impossibl